<a href="https://colab.research.google.com/github/arrazolahn/B1-Procesamiento-de-Datos-con-Python-2020-2021/blob/master/Proyecto_Bedu.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto BEDU - Procesamiento de datos con Python
Por Norma Arrazola Herrera
##1. Problema Identificado e Investigación:
Citando al sociologo Branko Milanovic 60 porciento del ingreso de una persona se determina por el país en que se nace, y luego 20 porciento adicional proviende de cuan ricos son los padres.
La pobreza es uno de los mayores problemas que afecta a la humanidad. El impacto de en cifras se puede resumir:
al entender que el 85% de la población mundial vive con $30 dolares al dia.

1.   10% de la población mundial vive en pobreza extrema.
2.   Aproximadamente la mitad de la humanidad vive con menos de $5.5 USD por dia

Las preguntas a resolver son:


1.   ¿Existe una relacion entre la geografia de un pais y su nivel de pobreza?
2.   ¿Cuanto influencia los recursos naturales de un pais y la pobreza de sus habitantes?
3. ¿Cuanto influencia el tamaño de un pais a su nievl de pobreza?
4. ¿Hay alguna relacion entre el tipo de clima predominante y el nivel de pobreza de un pais?
5. ¿Que tan importante es el terreno fertil para que sus pobladores no vivan en pobreza? 


##3. Soluciones hasta el momento
La mayoría de las soluciones de organismos como la ONU no analizan el problema desde una perspectiva historica y geografica lo que podria a ayudar a entender mejor el problema y proveer soluciones mas efectivas.

Fuentes

* https://www.worldvision.org/sponsorship-news-stories/global-poverty-facts#facts
* https://www.demos.org/blog/income-inequality-interview-branko-milanovic
* https://www.oxfam.org/en/5-shocking-facts-about-extreme-global-inequality-and-how-even-it

##4. Datasets a utilizar

World Bank. 2006. World Development Indicators. Washington, DC: World Bank.
https://diegopuga.org/data/rugged/#grid



In [1]:
import pandas as pd

url='https://raw.githubusercontent.com/arrazolahn/B1-Procesamiento-de-Datos-con-Python-2020-2021/master/Datasets/rugged_data.csv'
df = pd.read_csv(url,sep=",",error_bad_lines=False,encoding='latin-1')
print (df)
df.head()


dataset=df.loc[:, ['isocode','country', 'rgdppc_2000', 'land_area', 'lat','lon', 'soil', 'desert','tropical', 'dist_coast', 'near_coast', 'gemstones']]
print (dataset)
dataset.head()


    isocode  isonum  ...   pop_1400  european_descent
0       ABW     533  ...      614.0               NaN
1       AFG       4  ...  1870829.0               0.0
2       AGO      24  ...  1223208.0               2.0
3       AIA     660  ...        NaN               NaN
4       ALB       8  ...   200000.0             100.0
..      ...     ...  ...        ...               ...
229     WSM     882  ...    33072.0               NaN
230     YEM     887  ...  2250000.0               0.0
231     ZAF     710  ...   486527.0              18.0
232     ZMB     894  ...   348220.0               1.0
233     ZWE     716  ...   174110.0               1.0

[234 rows x 51 columns]
    isocode       country  rgdppc_2000  ...  dist_coast  near_coast  gemstones
0       ABW         Aruba          NaN  ...       0.001    100.0000          0
1       AFG   Afghanistan          NaN  ...       0.922      0.0000          0
2       AGO        Angola     1794.729  ...       0.428     13.1587      47756
3       AIA

isocode      country  rgdppc_2000  ...  dist_coast  near_coast  gemstones
0     ABW        Aruba          NaN  ...       0.001    100.0000          0
1     AFG  Afghanistan          NaN  ...       0.922      0.0000          0
2     AGO       Angola     1794.729  ...       0.428     13.1587      47756
3     AIA     Anguilla          NaN  ...       0.000    100.0000          0
4     ALB      Albania     3703.113  ...       0.048     94.6919          0

[5 rows x 12 columns]

In [2]:
dataset.describe()
dataset.loc[df['isocode'] == 'MEX']

isocode country  rgdppc_2000  ...  dist_coast  near_coast  gemstones
131     MEX  Mexico      9046.27  ...       0.183     38.4165          0

[1 rows x 12 columns]


##5. Revisión de Data Set
1. ¿El conjunto de datos que tengo realmente me sirve para responder algunas de las preguntas que me planteé?
2. ¿Qué tamaño tiene mi conjunto de datos? ¿Serán datos suficientes?
3. ¿Qué columnas tengo y qué información tengo en cada una de esas columnas?
4. Los nombres que tienen mis columnas, ¿son el nombre más apropiado?
5. ¿Qué tipos de datos tengo en cada columna? ¿Parecen ser el tipo correcto de datos? ¿O es un tipo de datos "incorrecto"?
6. Si selecciono algunas filas al azar y las observo, ¿estoy obteniendo los datos que debería? ¿o hay datos que parecen estar "sucios" o "incorrectos"?

Tenemos 234 paises con 12 datos, de los cuales 2 son indicadores (isocde y country), 1 es la variable dependiente (GDP) y 9 otras variables descibiendo la geografia del lugar son variables independientes. 

* country: Country name.
* land_area: Land area (1000 Ha). The source is the Food and Agriculture Organization (2008), except for Macau and Hong Kong where it is the Encyclopædia Britannica. Units are thousands of hectares.
* lat: Latitude. Expressed in decimal degress, for the geographical centroid of the country.
*lon: Longitude. Expressed in decimal degress, for the geographical centroid of the country.
* soil: % Fertile soil. On the basis of the FAO/UNESCO Digital Soil Map of the World and linked soil association composition table and climatic data compiled by the Climate Research Unit of the University of East Anglia, Fischer, van Velthuizen, Shah, and Nachtergaele (2002) identify whether each cell on a 5-minute grid covering almost the entire land area of the Earth is subject to various constraints for growing rainfed crops.
* tropical: % Tropical climate. 
* dist_coast: Average distance to nearest ice-free coast (1000 km).
* near_coast: % Within 100 km of ice-free coast. On the basis of the same data used to calculate the average distance to nearest ice-free coast, we calculate the percentage of the land surface area of each country that is within 100km of the nearest ice-free coast.
* gemstones: Gem diamond extraction 1958-2000 (1000 carats).  Units are thousands of carats.
* rgdppc_2000: Real GDP per person 2000 -- World Bank.  Units are 2006 international dollars, with purchasing power parity conversions performed using the Elteto-Koves-Szulc method.




##6. Limpieza de datos y preparación 

Explora tu dataset con el fin de encontrar los NaNs que contiene. Piensa en la distribución de NaNs por columna y por fila.
Piensa cuáles son los procedimientos que puedes aplicar a tus NaNs. 
¿Tenemos que eliminar las filas/columnas que tienen esos NaNs? ¿O podríamos rellenar esos NaNs con algún valor de manera que podamos retener esas filas/columnas?
Limpia tu dataset de manera que no quede ningún NaN.

*   Reindexa tu dataset si lo consideras necesario.
*   Renombra tus columnas si lo consideras necesario.
*   Prueba aplicar agregaciones a tu DataFrame para ver si puedes empezar a responder algunas de las preguntas que te planteaste anteriormente.
*   Elemento de la lista







In [3]:
dataset.isnull().sum().sort_values(ascending = False).head(12)

rgdppc_2000    64
soil            9
land_area       4
gemstones       0
near_coast      0
dist_coast      0
tropical        0
desert          0
lon             0
lat             0
country         0
isocode         0
dtype: int64

In [4]:
dataset[dataset['soil'].isnull()]

isocode                               country  ...  near_coast  gemstones
35      CCK               Cocos (Keeling) Islands  ...    100.0000          0
49      CXR                      Christmas Island  ...    100.0000          0
96      IOT        British Indian Ocean Territory  ...    100.0000          0
150     NFK                        Norfolk Island  ...    100.0000          0
157     NRU                                 Nauru  ...    100.0000          0
162     PCN                              Pitcairn  ...    100.0000          0
185     SJM        Svalbard and Jan Mayen Islands  ...     13.7615          0
205     TKL                               Tokelau  ...    100.0000          0
217     UMI  United States Minor Outlying Islands  ...    100.0000          0

[9 rows x 12 columns]

Todos los paises sin valor en soil son las que no tienen tampoco GDP. Motivo de esto puede ser que son colonias pertenecientes a otros paises (e.g. Chistmas Island pertenece a Australia).
Por ello eliminaremos estas columnas 

In [5]:
dataset[dataset['land_area'].isnull()]

isocode                               country  ...  near_coast  gemstones
127     MCO                                Monaco  ...    100.0000          0
185     SJM        Svalbard and Jan Mayen Islands  ...     13.7615          0
217     UMI  United States Minor Outlying Islands  ...    100.0000          0
221     VAT                              Holy See  ...    100.0000          0

[4 rows x 12 columns]

Monaco es el unico pais reconocido de la lista, ademas de ser el segundo pais mas pequeño en el mundo con un area de  2.1 km2, lo cual tiende a cero usando nuestra metodología (unidad en miles de hectareas)

In [8]:
print(dataset[dataset['rgdppc_2000'].isnull()])

    isocode                               country  ...  near_coast  gemstones
0       ABW                                 Aruba  ...       100.0          0
1       AFG                           Afghanistan  ...         0.0          0
3       AIA                              Anguilla  ...       100.0          0
5       AND                               Andorra  ...         0.0          0
6       ANT                  Netherlands Antilles  ...       100.0          0
..      ...                                   ...  ...         ...        ...
217     UMI  United States Minor Outlying Islands  ...       100.0          0
221     VAT                              Holy See  ...       100.0          0
224     VGB                British Virgin Islands  ...       100.0          0
225     VIR          United States Virgin Islands  ...       100.0          0
228     WLF             Wallis and Futuna Islands  ...       100.0          0

[64 rows x 12 columns]


De neuvo vemos que muchas islas y colonias (e.g. Aruba) son las que no cuentan con nuestra variable independiente por lo que procedemos a eliminar estas filas.

In [11]:
finaldataset = dataset.dropna(subset=["rgdppc_2000"])
print (finaldataset)
finaldataset.isnull().sum().sort_values(ascending = False).head(12)

    isocode               country  ...  near_coast  gemstones
2       AGO                Angola  ...     13.1587      47756
4       ALB               Albania  ...     94.6919          0
7       ARE  United Arab Emirates  ...     75.7464          0
8       ARG             Argentina  ...     13.0167          0
9       ARM               Armenia  ...      0.0000          0
..      ...                   ...  ...         ...        ...
229     WSM                 Samoa  ...    100.0000          0
230     YEM                 Yemen  ...     38.1163          0
231     ZAF          South Africa  ...     18.9404     144785
232     ZMB                Zambia  ...      0.0000          0
233     ZWE              Zimbabwe  ...      0.0000        898

[170 rows x 12 columns]


gemstones      0
near_coast     0
dist_coast     0
tropical       0
desert         0
soil           0
lon            0
lat            0
land_area      0
rgdppc_2000    0
country        0
isocode        0
dtype: int64

Terminamos con un dataset con todas las variables dependientes e independientes completas, habiendo identificado posibles razones de porque habia valores perdidos.

##7. Uso de APIs
En este Postwork vamos a dejar descansar un poquito nuestro dataset. La única manera de entender a la perfección cómo utilizar APIs es usándolas. Y eso es lo que vamos a hacer. Vamos a elegir un API y construir un nuevo dataset utilizando peticiones HTTP y concatenaciones. Puede que sea un API que complemente tu dataset original (eso sería lo mejor de lo mejor), o puede simplemente que sea un API que ofrezca información que te parezca interesante. Los pasos a seguir son estos:



1. Encuentra un API que quieras explorar. Puedes encontrar una lista enorme de APIs gratuitas [aquí](https://github.com/public-apis/public-apis).
2. Crea una cuenta si es necesario.
3. Lee la documentación.
4. Realiza algunas peticiones de prueba para entender la estructura de los datos (si quieres explorar un poco, puedes intentar haacer peticiones usando [este software](https://www.postman.com/)).
5. Automatiza el proceso de realizar peticiones para obtener un dataset considerablemente grande.
6. Explora y limpia tu dataset.
7. Si has encontrado un dataset que complementa la información de tu dataset original (el que has estado trabajando durante todas las sesiones), ve si puedes unirlos en un solo `DataFrame` que incluya la información útil de ambos.

> **Extra**: Puedes también explorar la posibilidad de automatizar algunos de los procesos de limpieza que has realizado con tu dataset. Si encuentras dónde podría ser útil y te animas a hacerlo, recuerda proteger tu código con estructuras `try except` para que la automatización no falle.


https://datacatalog.worldbank.org/dataset/poverty-and-equity-database

https://datahelpdesk.worldbank.org/knowledgebase/topics/125589

https://datahelpdesk.worldbank.org/knowledgebase/articles/898581

http://wdi.worldbank.org/table/3.14


In [12]:
#!pip install requests
import pandas as pd
import requests
import time


endpoint = "https://api.worldbank.org/v2/country/all/indicator/SI.POV.DDAY?format=json&date=2018"
r=requests.get(endpoint).json()
print (r)
type (r)

#r.status_code #revision de la respuesta
#r.json() #examinar el json obtenido
#r.keys()

[{'page': 1, 'pages': 6, 'per_page': 50, 'total': 266, 'sourceid': '2', 'sourcename': 'World Development Indicators', 'lastupdated': '2021-07-30'}, [{'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'}, 'countryiso3code': 'AFE', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}, {'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': 'ZI', 'value': 'Africa Western and Central'}, 'countryiso3code': 'AFW', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}, {'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': '1A', 'value': 'Arab World'}, 'countryiso3code': 'ARB', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}, {'indicator': {'i

list

In [18]:
type(r) #list
response=r[1]
#type(response) #list
print(response)
len(response) #50

[{'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'}, 'countryiso3code': 'AFE', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}, {'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': 'ZI', 'value': 'Africa Western and Central'}, 'countryiso3code': 'AFW', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}, {'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': '1A', 'value': 'Arab World'}, 'countryiso3code': 'ARB', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}, {'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': 'S3', 'value': 'Caribbean smal

50

In [16]:
print(response[0])
len(response[0]) #8

{'indicator': {'id': 'SI.POV.DDAY', 'value': 'Poverty headcount ratio at $1.90 a day (2011 PPP) (% of population)'}, 'country': {'id': 'ZH', 'value': 'Africa Eastern and Southern'}, 'countryiso3code': 'AFE', 'date': '2018', 'value': None, 'unit': '', 'obs_status': '', 'decimal': 1}


8

In [19]:
#obtener el pais
type(response[0]) #dict
#print(response[0]["country"]["id"])

rows = []
len(response[0])
x = range(51)
for i in x:
  print(response[i]["countryiso3code"]) #obtener el codigo Iso del pais
  print(response[i]["value"]) #obtener el valor de % de gente viviendo con 1.9 dls al dia
  


AFE
None
AFW
None
ARB
None
CSS
None
CEB
None
EAR
None
EAS
1.2
EAP
None
TEA
None
EMU
None
ECS
1.1
ECA
None
TEC
None
EUU
None
FCS
None
HPC
None

0.6
IBD
None
IBT
None
IDB
None
IDX
None
IDA
29.3
LTE
None
LCN
3.7
LAC
None
TLA
None
LDC
None
LMY
None

None

12.8
MEA
7
MNA
None
TMN
None
MIC
None
NAC
None

None
OED
None
OSS
None
PSS
None
PST
None
PRE
None
SST
None
SAS
None
TSA
None
SSF
40.4
SSA
None
TSS
None

1.4
WLD
None
AFG
None


IndexError: ignored

Devido al alto numero de datos vacios y la calidad de la información se decidió por no incluir esta variable en el analisis.

##8. Procesamiento de Datos
Ha llegado el momento de aplicar técnicas comunes de procesamiento de datos a nuestro conjunto de datos. Para este momento ya debes de tener un conjunto de datos que te sea interesante y que haya sido limpiado previamente de `NaNs`. También es importante que ya hayas explorado extensivamente tu dataset para que lo conozcas muy bien. De igual manera, es recomendable que tu dataset esté bien indexado y que tenga nombres coherentes y comprensibles para las columnas.

Para aplicar en tu dataset lo que vimos en esta sesión, realiza las siguientes acciones:

1. Checa que todos tus datos tengan el tipo de dato correcto. Si no es así, usa casting para convertir tus datos al tipo de dato correcto (recuerda que tipos de dato como `datetime64` se guardan como strings cuando están en archivos .csv, así que tendrás que convertirlos al tipo de dato apropiado cada vez que importes tu archivo.)
2. Si tienes columnas de texto, asegúrate de que todas tengan el formato correcto. Si no es así, utiliza las técnicas de manipulación de `strings` para darles el formato que necesitas.
3. Si consideras que alguna de tus columnas sería más clara si los datos tuvieran otro formato o representación usa `map` para transformar los datos de esa columna.
4. Si crees que es posible generar nuevas columnas útiles a partir de las columnas que ya tienes, usa `apply` para generar nuevos datos a partir de los que tienes y añádelos a tu dataset.
5. Con el fin de responder algunas de las preguntas que te planteaste acerca de tu dataset, usa filtros y sorting para crear nuevos subconjuntos y reordenamientos que sean más adecuados para responder tus preguntas. Primero comienza intentando responder las preguntas que te planteaste al principio, pero después puedes solamente explorar para ver si encuentras otras preguntas que no te habías planteado anteriormente.

##9. Planeación estatistica.

¿Qué te gustaría analizar? ¿Qué información está disponible en tus datos? ¿Cómo podrías ayudarte a entender mejor los datos usando gráficas y visualizaciones? Has una lista de tus planes a futuro. Seguramente esa lista irá cambiando conforme avances en el siguiente módulo, pero es un gran comienzo para tener una dirección clara.

* Regresiones lineales
* Correlaciones